### ISSUE TITLE
[EPCO-25] Training will be egoized

### ISSUE THREAD

**Issue by hatfield-c** [hatfield-c/epco#25]
Part of the appeal of the current EPCO methodology is that it goes around the issue of dimensionality explosion. The trained model does not learn a mapping of every input to every action and all the possible outputs, as doing so increases dimensionality issues far too much under the current architecture.

This presents a problem, as the current valid/invalid differentiation is done across the entire domain of the problem space, *however* there is no encoding of what actions could lead to what resulting states. This precludes the possibility of planning, as there would be no way to build a coherent plan without some sort of onboard simulation.

Egoization is thus proposed as an alternative to learning a mapping between all possible system states to actions to resulting states. The model is constrained to only learn what the next set of possible system states is given the current state for any action taken. Thus we can discretize the search process into a series of jumps from a position in one set of reachable states to the next position in another set of reachable states.

However. this new paradigm presents an issue as well. How do we generate negative examples? The generation of explicitly positive examples is very easy - simply take random actions around the environment and observe whats possible. However it remains to be seen how negative examples can be generated.

**Comment by hatfield-c** [comment]
The typical state is:
- [ x_pos, y_pos, x_v, y_v, theta, theta_v, Right, Left ]

To egoize the model, we remove global context from its observations. Thus the first two positions in the state are understood to be the resulting x_pos and y_pos differentials caused by taking an action. It is up to the global planner to navigate the global state by egoizing its data before feeding it to the VNN, and then de-egoizing data from the VNN.

**Comment by hatfield-c** [comment]
There is an issue presented however, by the fact that there are global constraints on position. Typically this would be addressed via lidar detection, however there is no such devices available in the lunar lander scenario. Thus how can the planner enforce these constraints on position.

This speaks to a need for two levels of planning. One for these static, global constraints, and the other for encoding what states are possible to reach in the next frame and what action will help reach it. Thus the global planner uses the ego constraints to determine what is immediately possible, and picks points from the localized constraint space. These points are then chained together to make global plans. During the process where a point is picked, the ego constraints are shifted via egoization such that it's origin lies at the point of interest, and then the next planning step is performed.

**Comment by hatfield-c** [comment]
The action space component of the planner works by informing the planner of what space is reachable by the agent within a single time step. However, it also must encode what action was taken to reach that state, otherwise the planner won't be able to command the agent. This is achieved by training a VNN both on the world space as well as on additional variables that describe how the world's parameters is changing. In a way, the key to success involves this mapping of actions and the environment's state to these gradients. These outputs on the changing values can easily be used to inform the planner of what next states are possible, however the process of generating them doesn't need to be voxelized.

Indeed, potentially any method can be used to generate the gradient outputs. A neural network or event an existing flight controller algorithm can be used. The issue with these approaches however is they are not interpretable, so they cannot easily tell us the entire set of possible choices.

**Comment by hatfield-c** [comment]
The simplest way to set up the egoizied gradient outputs is to simply have a gradient for every variable. Thus the entire space becomes:
- [ x, y, theta, L, R, x', y', theta', x'', y'', theta'', b1, b2 ]

It should be possible to optimize this even further to remove redudant relationships, i.e. x'' -> x' -> x. For now focus on keeping this simple.

I need to more explicitly define how unification occurs under this paradigm before moving forward.

**Comment by hatfield-c** [comment]
We are interested in selecting the "next state" that occurs after an action for examination. This can be calculated relatively easily by adding the gradient of each variable to itself. 

What we are interested in is a set of voxels in world space that gives us the set of immediately reachable states. This information must be extracted from the action space, which can be done through the following process:
1. Assume you are given a state from the action space
2. Take the subset of the action space voxels that corresponds to the world space (that is to say, project the action voxels into world space by dropping the extra linear separators).
3. Gather the voxels which contain the world space subset values that can extracted from the action sample
4. Choose these corresponding voxels from the original action space.
5. Drop the world space subset from these newly voxels
6. Add each offset value from each linear separator of the gradients to the corresponding world space variable they are a gradient to.
7. Create a new voxel in world space for each voxel in the action space, and make the offset of the linear separator equal to the value calculated in the previous step.
8. You now have the set of voxels in real space that give the immediately reachable next set of states. 

We'll need to keep track of the ID's of the voxels during this process. At the end, when the next state is selected, the originating action should be recorded as well. The action can be extracted by using the chosen voxel ID to select a voxel from action space, and then using a heuristic to choose an action from the voxel. For now, I'll use a simple heuristic of choosing an action value that lies between the linear separators that control the action dimension.

**Comment by hatfield-c** [comment]
What happens if multiple voxels are chosen (i.e. they overlap)?
- Choose the one with the smallest difference between linear separators in the action dimension.

**Comment by hatfield-c** [comment]
This issue will unfortunately need to be cancelled, as the lunar lander environment is not suitable for our training purposes. 
The lunar lander environment cannot set an initial state other than the default it comes with, and my attempts to hack around this limitation has been met with failure.

As such, I will transition to the Mujoco environment, as it can set an initial state.
